src/gui/MainWindow.cpp

```diff
@@ -1281,40 +1279,46 @@ void MainWindow::updateTrayIcon()
             menu->addAction(m_ui->actionLockDatabases);
 
 #ifdef Q_OS_MACOS
-            QAction* actionQuit = new QAction(tr("Quit KeePassXC"), menu);
-            menu->addAction(actionQuit);
-
+            auto actionQuit = new QAction(tr("Quit KeePassXC"), menu);
             connect(actionQuit, SIGNAL(triggered()), SLOT(appExit()));
+            menu->addAction(actionQuit);
 #else
             menu->addAction(m_ui->actionQuit);
-
 #endif
+            m_trayIcon->setContextMenu(menu);
+
             connect(m_trayIcon,
                     SIGNAL(activated(QSystemTrayIcon::ActivationReason)),
                     SLOT(trayIconTriggered(QSystemTrayIcon::ActivationReason)));
             connect(actionToggle, SIGNAL(triggered()), SLOT(toggleWindow()));
-
-            m_trayIcon->setContextMenu(menu);
-
-            m_trayIcon->setIcon(resources()->trayIcon());
-            m_trayIcon->show();
         }
 
-        if (m_ui->tabWidget->count() == 0) {
-            m_trayIcon->setIcon(resources()->trayIcon());
-        } else if (m_ui->tabWidget->hasLockableDatabases()) {
+        if (m_ui->tabWidget->hasLockableDatabases()) {
             m_trayIcon->setIcon(resources()->trayIconUnlocked());
         } else {
             m_trayIcon->setIcon(resources()->trayIconLocked());
         }
-    } else {
-        QApplication::setQuitOnLastWindowClosed(true);
 
+        m_trayIcon->setToolTip(windowTitle().replace("[*]", isWindowModified() ? "*" : ""));
+        m_trayIcon->show();
+
+        if (!isTrayIconEnabled() || !QSystemTrayIcon::isSystemTrayAvailable()) {
+            // Try to show tray icon after 5 seconds, try 5 times
+            // This can happen if KeePassXC starts before the system tray is available
```

Comment:
I wonder if there are DBus signals we can listen for.

Reply:
That is Qt's job which it is failing at doing (that's actually their bug as reported). This works and isn't too bothersome.

Reply:
It's definitely Qt's job to do it right, but we might find a more elegant workaround in the future perhaps.

Reply:
Is agree to that